### PR TITLE
Rewrite JS3 week 1 prep

### PR DESCRIPTION
**Pull request by Dedekind561**
## What does this change?

Module: JS3 
Week(s): 1

## Checklist

- [x] I have read the [contributing guidelines](CONTRIBUTING.MD)
- [x] I have checked my spelling and grammar with an [automated tool](https://www.grammarly.com/grammar-check)
- [x] I have previewed my changes to check the [markdown renders](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) as I intend
- [x] I have run my code to check it works
- [x] My changes follow our [Style Guide](https://curriculum.codeyourfuture.io/guides/code-style-guide)

### CODE REVIEW

**Comment by Dedekind561** [comment]
I'm going to continue crafting this prep section over the next few days.

File: content/js3/blocks/call-stack/index.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+title = '🗼 Call stack'
+headless = true
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Explain how data is rendered into a user interface'
++++
+
+When playing computer, we can see that when we call `map` then the function `double` is also invoked too. So whilst one function is called we make a call to another function and then \*_return_ back to the previous frame. In other words, we need to keep track of the which function is currently executing.
```
Comment (SallyMcGrath) [inline content/js3/blocks/call-stack/index.md:11]:
```suggestion
When playing computer, we can see that when we call `map` then the function `double` is also invoked. While one function is called, we make a call to another function and then _return_ to the previous frame. In other words, we need to keep track of which function is currently executing.
```

File: content/js3/blocks/components/index.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+const section = document.createElement("section");
+section.innerHTML = `
+<p>${show.name}</p>
+<p>${show.date}</p>
+<p>${show.location}</p>
+`;
+console.log(section);
+```
+
+Our sub-goal is to be able to render _any_ show object in the DOM. To do this, we can wrap up create a show card component using a function.
```
Comment (SallyMcGrath) [inline content/js3/blocks/components/index.md:32]:
```suggestion
Our sub-goal is to be able to render _any_ "show" object in the DOM. To do this, we can create a card component using a function.
```

File: content/js3/blocks/components/index.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  city: "Manchester",
+};
+const show2 = {
+  name: "Gang of Four",
+  date: "5th November",
+  location: "02 Victoria Warehouse",
+  city: "Manchester",
+};
+
+function createShowCard() {
```
Comment (SallyMcGrath) [inline content/js3/blocks/components/index.md:48]:
```suggestion
function createShowCard() {
```
```suggestion
function createCard(show) {
```

File: content/js3/blocks/components/index.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  const section = document.createElement("section");
+  section.innerHTML = `
+    <p>${show.name}</p>
+    <p>${show.date}</p>
+    <p>${show.location}</p>
+`;
+  return section;
+}
+
+document.append(createShowCard(show1), createShowCard(show2)); // append the show cards to the DOM
```
Comment (SallyMcGrath) [inline content/js3/blocks/components/index.md:58]:
```suggestion
document.append(createCard(show1), createCard(show2)); // append the cards to the DOM
```

File: content/js3/blocks/components/index.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Explain how data is rendered into a user interface'
++++
+
+To build user interfaces, we break down the interface into smaller **components**. Functions are reusable blocks of code that we can use to do the same thing again and again. In a similar way, we build a complex user interface from **components** - resuable blocks of the user interface. As functions allow us to reuse code again and again, we can implement components using functions.
+
+Let's consider our code thus far:
```
Comment (SallyMcGrath) [inline content/js3/blocks/components/index.md:13]:
```suggestion
Look at our code so far:
```

File: content/js3/blocks/components/index.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+title = '🧩 Components'
+headless = true
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Explain how data is rendered into a user interface'
++++
+
+To build user interfaces, we break down the interface into smaller **components**. Functions are reusable blocks of code that we can use to do the same thing again and again. In a similar way, we build a complex user interface from **components** - resuable blocks of the user interface. As functions allow us to reuse code again and again, we can implement components using functions.
```
Comment (SallyMcGrath) [inline content/js3/blocks/components/index.md:11]:
```suggestion
We already know that to build user interfaces, we break down the interface into smaller **components**. 

Functions are reusable blocks of code that we can use to do the same thing again and again. In a similar way, we build a complex user interface from **components**. Components are reusable blocks of the user interface. As functions allow us to reuse code again and again, we can implement components using functions.
```

File: content/js3/blocks/components/index.md (hunk 2 above)
Comment (SallyMcGrath) [reply content/js3/blocks/components/index.md:32]: I just find the show card a bit hard to read as I keep reading it as showing the card, so I've made it card and passed show in.

File: content/js3/blocks/data-ui/index.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+title = '💾 ➡️ 💻 Data to UI '
+headless = true
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Explain how data is rendered into a user interface'
++++
+
+When we're building user interfaces we're often taking data and {{<tooltip title="rendering">}}**rendering** is the process of building a visual representation or interface from some code{{</tooltip>}} it in the user interface. We're going to start with some data and explore how this data structure is rendered on the page.
```
Comment (SallyMcGrath) [inline content/js3/blocks/data-ui/index.md:11]:
```suggestion
User interfaces render data. {{<tooltip title="rendering">}}**rendering** is the process of building a visual representation or interface from some code{{</tooltip>}} We're going to start with some data and explore how this data structure is rendered on the page.
```

File: content/js3/blocks/data-ui/index.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+### ℹ️ Show listings
+
+Suppose you're building a user interface to display information about local shows in a city.
+
+Let's consider the following problem:
```
Comment (SallyMcGrath) [inline content/js3/blocks/data-ui/index.md:17]:
```suggestion

```

File: content/js3/blocks/data-ui/index.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+> _Given_ a list of shows data
+> _When_ the page first loads
+> _Then_ it should display the list of events, including the data of the name, date and location of the event
+
+A typical event can be modelled as a JavaScript object with the following properties:
```
Comment (SallyMcGrath) [inline content/js3/blocks/data-ui/index.md:23]:
```suggestion
A show can be modelled as a JavaScript object with the following properties:
```

An event can be mistaken for an event in a program.

File: content/js3/blocks/data-ui/index.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+// js object with event data
+{
+  name: "Little Simz",
+  date: "5th November",
+  location: "02 Victoria Warehouse",
+  city: "Manchester"
+}
+```
+
+So our list of event data can be modelled as an array of objects, each with a similar shape:
```
Comment (SallyMcGrath) [inline content/js3/blocks/data-ui/index.md:35]:
```suggestion
So our list of show data can be modelled as an array of objects, each with a similar shape:
```

If we're saying show, let's always say it wrt the data.

File: content/js3/blocks/data-ui/index.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+```js
+const events = [
```
Comment (SallyMcGrath) [inline content/js3/blocks/data-ui/index.md:38]:
```suggestion
const shows = [
```

File: content/js3/blocks/multiple-data/index.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Use the Teach Tech Together guide to construct your objectives'
+    2='Limit the objectives to 3-5 items'
+    3='Write objectives you can measure'
++++
+
+We can take a single datum and render it in the UI. However, we will often have larger data structures that hold collections of data that we want to render in the user interface. Let's consider the folllowing:
```
Comment (SallyMcGrath) [inline content/js3/blocks/multiple-data/index.md:13]:
```suggestion
A datum is a single piece of data. We can take a single datum and render it in the UI. More often we render collections of data in the user interface.
```

File: content/js3/blocks/multiple-data/index.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+const cards = [];
+for (const item of shows) {
+  cards.push(createShowCard(item));
+}
+
+document.querySelector("ul").append(...elements);
+// invoke append using the spread operator
+```
+
+We want to map over this data and render it in the user interface.
```
Comment (SallyMcGrath) [inline content/js3/blocks/multiple-data/index.md:42]:
```suggestion
We want to loop over this data and render it in the user interface.
```
You've said map but not used map yet. Do you want to loop here?

File: content/js3/blocks/one-to-one/index.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+title = 'One to one'
+headless = true
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Define a one to one mapping'
++++
+
+At the moment, we have a one to one mapping between the array of data and the UI elements. We can take a group of UI elements and
```
Comment (SallyMcGrath) [inline content/js3/blocks/one-to-one/index.md:11]:
```suggestion
At the moment, we have a one to one mapping between the array of data and the UI elements. Each item in the array matches a node in the UI.
```

File: content/js3/blocks/single-datum/index.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+title = '💽 Single datum'
+headless = true
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Describe how to render a single item of data in the browser'
++++
+
+To break down this problem, let's define a sub-goal:
```
Comment (SallyMcGrath) [inline content/js3/blocks/single-datum/index.md:11]:
```suggestion
```

File: content/js3/blocks/single-datum/index.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+> 🎯 Sub-goal: Render a single datum in the user interface
+
+We'll consider how to render a single datum before doing this for the whole list. We will start with an object like the one below:
```
Comment (SallyMcGrath) [inline content/js3/blocks/single-datum/index.md:15]:
```suggestion
To break down this problem, we'll render a single datum, before doing this for the whole list. Here's one show:
```

File: content/js3/blocks/using-map/index.md [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    1='Describe how to use map'
+    2='Describe why we use map instead of other iteration techniques'
++++
+
+### Using `map`
+
+For every item in the starting array we want to apply a function to all the elements in the array to create a new array.
+Earlier, we used a `for...of` statement to apply the function `createShowCard` to each element in the array. However, we can also use `map` to do this.
+
+`map` is a {{<tooltip title="higher order function">}} A **higher order function** is a function that takes another function as an argument or returns a new function{{</tooltip>}}. In this case, it means we pass a function as an argument to `map`. Then `map` will use this function to create a new array. Let's consider an example with a list of numbers and a function `Math.round` that returns the rounded number. Our goal is to create a new array of rounded numbers from the starting array of numbers:
```
Comment (SallyMcGrath) [inline content/js3/blocks/using-map/index.md:17]:
```suggestion
`map` is a {{<tooltip title="higher order function">}} A **higher order function** is a function that takes another function as an argument or returns a new function{{</tooltip>}}. In this case, it means we pass a function as an argument to `map`. Then `map` will use this function to create a new array. 

Consider a list of numbers and a function `Math.round` that returns the rounded number. Our goal is to create a new array of rounded numbers from the starting array of numbers:
```

**Review by SallyMcGrath — CHANGES_REQUESTED** [review]
Mini copy edit -- I'm just here to get the details for the workshop (components-2)

File: content/js3/blocks/single-datum/index.md [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
++++
+
+> 🎯 Sub-goal: Build a film card component
+
+To break down this problem, we'll render a single datum, before doing this for the whole list. Here's one film:
+
+```js
+const film = {
+    title: "Killing of Flower Moon",
+    director: "Martin Scoresese"
```
Comment (SallyMcGrath) [inline content/js3/blocks/single-datum/index.md:18]:
```suggestion
    director: "Martin Scorsese"
```

File: content/js3/blocks/using-map/index.md [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+```js
+const arr = [5, 20, 30];
+
+function double(num) {
+  return num * 2;
+}
+```
+
+In this example, our goal is to create a new array of doubled numbers given this array and function. In other words, we want to create the array `[10, 40, 60]`. Firstly, we can observe a one-to-one mapping:
```
Comment (SallyMcGrath) [inline content/js3/blocks/using-map/index.md:23]:
```suggestion
Our goal is to create a new array of doubled numbers given this array and function. We want to create the array `[10, 40, 60]`.  Look, it's another "one to one mapping"
```

File: content/js3/blocks/using-map/index.md [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+{{<tab name="🔨Implement">}}
+
+Given the list of film data:
+
+```js
+const films = [
+  {
+    title: "Killing of Flower Moon",
+    director: "Martin Scoresee"
```
Comment (SallyMcGrath) [inline content/js3/blocks/using-map/index.md:89]:
```suggestion
    director: "Martin Scorsese"
```

File: content/js3/blocks/using-map/index.md [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+    times: ["15:35"],
+    certificate: "15",
+    duration: 112
+  },
+  {
+    title: "Typist Artist Pirate King",
+    directory: "Carol Morley"
```
Comment (SallyMcGrath) [inline content/js3/blocks/using-map/index.md:96]:
```suggestion
    director: "Carol Morley"
```

**Review by SallyMcGrath — CHANGES_REQUESTED** [review]
minor revisions

File: assets/styles/03-elements/images.scss [hunk 22]
```diff
@@ -18,6 +18,7 @@ picture {
 figure {
   width: fit-content;
   box-shadow: var(--theme-box-shadow);
+  margin: 0;
```
[the remark this answers is not in the document]
Comment (Dedekind561) [reply assets/styles/03-elements/images.scss:21]: Thanks for catching this.

**Pull request merged**